presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java

```diff
@@ -611,6 +614,57 @@ public void finishDelete(Session session, TableHandle tableHandle, Collection<Sl
     }
 
     @Override
+    public TableHandle beginSelect(Session session, TableHandle tableHandle, Optional<TableLayoutHandle> layoutHandle, Collection<ColumnHandle> columnHandles)
+    {
+        ConnectorEntry entry = lookupConnectorFor(tableHandle);
+        ConnectorMetadata metadata = entry.getMetadata(session);
+        ConnectorTableHandle connectorHandle = metadata.beginSelect(session.toConnectorSession(entry.getCatalog()), tableHandle.getConnectorHandle(), convertToConnectorTableLayoutHandle(layoutHandle), columnHandles);
+        if (connectorHandle == null) {
+            return null;
+        }
+        TableHandle th = new TableHandle(tableHandle.getConnectorId(), connectorHandle);
+        registerTableForQueryId(session.getQueryId(), th);
+        return th;
+    }
+
+    private void registerTableForQueryId(QueryId queryId, TableHandle tableHandle)
+    {
+        tablesToNotifyByQueryId.putIfAbsent(queryId.getId(), new ArrayList<>());
+        List<TableHandle> tables = tablesToNotifyByQueryId.get(queryId.getId());
+        tables.add(tableHandle);
+    }
+
+    @Override
+    public void finishSelect(Session session)
+    {
+        String queryId = session.getQueryId().getId();
+        try {
+            List<TableHandle> tables = tablesToNotifyByQueryId.get(queryId);
```

Comment:
This and 3 lines below could be moved outside `try-finally` block to avoid calling https://github.com/prestodb/presto/pull/5122/files#diff-dfecb0b24d9d5c723d964cb59873b44eR653 when there are no tables to notify but I kept it for consistency inside. @electrum any preferences? Is performance impact meaningful in this part?